pkg/task/task.go

```diff
@@ -0,0 +1,68 @@
+package task
+
+import "time"
+
+// TaskState (kind: int) represents the last known state of a task.
+// A task can be in one of three states
+// TaskStateScheduled: this is the initial state of the task when it enters the queue.
+// TaskStateProcessing: once work begins on the task, it is put into this state.
+// TaskStateComplete: work is no longer being done on this task. client should check task result.
+type TaskState int
+
+const (
+	StateScheduled TaskState = iota
+	StateProcessing
+	StateComplete
+)
+
+func (t TaskState) String() string {
+	return [...]string{
+		"StateRequested",
+		"StateProcessing",
+		"StateComplete",
+	}[t]
+}
+
+// TaskResultStatus (kind: int) is a status code for completed tasks.
+// TaskResultNone: initial status, No status, probably becasue the task is incomplete.
+// TaskResultSuccess: the task has completed without an error.
+// TaskResultFail: the task has completed with a failure.
+// TaskResultAbort: testground encountered an error and the task has not been scheduled.
+// testground will respond to client with a TaskResultAbort status if it is unable to enqueue
+// the reqeusted task.
+type TaskResultStatus int
+
+const (
+	ResultNone TaskResultStatus = iota
+	ResultSuccess
+	ResultFail
+	ResultAbort
+)
+
+func (t TaskResultStatus) String() string {
+	return [...]string{
+		"ResultNone",
+		"ResultSuccess",
+		"ResultFail",
+		"ResultAbort",
+	}[t]
+}
+
+// TaskResult (kind: struct)  contains a status code. If the status is not TaskResultNone or
+// TaskResultSuccess, relevant errors will be included in this struct.
+type TaskResult struct {
+	Status TaskResultStatus `json:"status"`
+	Errors []error          `json:"errors"`
```

Comment:
This is something we have to discuss on how to implement, and whether we need at the moment - for example if a test run fails, because out of 100 pods, 1 went out-of-memory, currently we don't detect that, and what would happen is that the test run will reach the global timeout and be killed (i.e. the way I understand the results above, we would just get a `ResultFail`, and upon inspecting the `outputs` we would notice that 99 have a context deadline exceeded, and 1 is killed).

Ideally we need to improve on this.

Reply:
On a second thought - we definitely want to propagate `events` or `statuses` from Kubernetes to Testground, because in TaaS developers won't have access to the cluster, all their interactions would go though Testground.

Therefore if out of 1000 pods:
997: Complete successfully
1: Panic because of testplan code
1: Fails due to pulling image
1: Fails due to OOM

I imagine we want to propagate all this information back to the user, so we should model it under the schema that will track historical state of jobs.